WinHTTP decoding

EmmaStott opened the pull request:
Added support for decoding the machine's current WinHTTP settings. The output will be written to winhttp_decoded.txt in the respective DC's folder.

# Notes
- Version string unchanged; this can be changed at your discretion.
- GitHub shows changes to line numbers that are untrue. ¯\_(ツ)_/¯

EmmaStott commented:
@googlebot I signed it!

eesheesh commented on gapstool.vbs:
```diff
@@ -10,17 +10,24 @@
 ' Do not change this line's format, build.bat relies on it.
-Const Ver = "1.0.1.0"

+Const Ver = "1.0.1.0"
```
Why is this showing as changed? This and other places. Maybe your branch is out of date (not synced to master or whatnot)?

EmmaStott replied:
I have no idea. The only thing I can think of is that the line numbers changed slightly due to CR/LF issues. I tried saving my changes using Visual Studio Code, SublimeText, and Notepad, but they Git always showed these lines as being altered.

I'll take another look this evening.

eesheesh replied:
I think we found that this was happening because the newlines were inconsistent, right?

eesheesh commented on gapstool.vbs:
```diff
@@ -306,14 +313,70 @@ Sub RunDiagnostics(CompName)
              "winhttp"
 
+  PrintLine "Getting system-wide WinHTTP settings dump from registry, and decoding - winhttp_decoded.txt"
+  ' Create a WMI StdRegProv.
```
This (until line 378, probably) is a large chunk, please put it in a Sub.

eesheesh commented on gapstool.vbs:
```diff
@@ -10,17 +10,24 @@
 Set objShell = WScript.CreateObject("Wscript.Shell")
 Const ForReading = 1, ForWriting = 2, ForAppending = 8
 Const WshRunning = 0, WshFinished = 1, WshFailed = 2
+Const HKEY_CLASSES_ROOT   = &H80000000
```
If you're only using HKLM, no need to add the other consts.

eesheesh commented on gapstool.vbs:
```diff
@@ -306,14 +313,70 @@ Sub RunDiagnostics(CompName)
+  PrintErrorIfNeeded "Error opening WMI StdRegProv on " & CompName & ": "
+  ' Retrieve the value of WinHTTPSettings from the registry.
+  ' Note that GetBinaryValue returns an array, where each element in the array
+  ' is a DECIMAL value of the octets. ¯\_(ツ)_/¯
```
Please avoid Unicode characters, I'm not sure what effects they'll have on weird old systems. :)

eesheesh commented on gapstool.vbs:
```diff
@@ -306,14 +313,70 @@ Sub RunDiagnostics(CompName)
+  ' is a DECIMAL value of the octets. ¯\_(ツ)_/¯
+  Dim WinHTTPSettingsArray
+  objStdRegProv.GetBinaryValue HKEY_LOCAL_MACHINE, _
+                              "SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\Connections", _
```
Please align the arguments.

eesheesh commented on gapstool.vbs:
```diff
@@ -306,14 +313,70 @@ Sub RunDiagnostics(CompName)
+                              "WinHttpSettings", _
+                              WinHTTPSettingsArray
+  PrintErrorIfNeeded "Error retrieving HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\Connections\WinHttpSettings on " & CompName & ": "
+  ' The WinHttpSettings registry value appears to be formatted as follows:
```
Is there any documentation about this?

EmmaStott replied:
I wasn't able to find any documentation, and I basically worked it out by looking at how https://p0w3rsh3ll.wordpress.com/2012/10/07/getsetclear-proxy/ was stepping through the value and determining how long each section was.

Magic numbers are bad, but the only other way I can think of to get this information would be to run `netsh winhttp show proxy` on the target machine and capture the response.

EmmaStott commented:
Requested changes have been made in a0d0af0223b479179dc1317feb3002969d98f1dc:
- Resolved CR/LF issues on a few lines. All line endings are now CRLF.
- Removed unnecessary registry constants since we're only using HKLM for now.
- Removed unicode shrug to prevent any parsing errors.
- Aligned the arguments of `objStdRegProv.GetBinaryValue`.
- Moved the new block of code into it's own sub, `DecodeWinHTTPSettings`. The name of the computer and the output file path are passed as arguments.

eesheesh commented on gapstool.vbs:
```diff
@@ -306,14 +368,17 @@ Sub RunDiagnostics(CompName)
              "winhttp"
 
+  PrintLine "Getting system-wide WinHTTP settings dump from registry, and decoding - winhttp_decoded.txt"
+  DecodeWinHTTPSettings CompName, LogDir & "\winhttp_decoded.txt"
```
You should save the file to `winhttp_decoded.txt` instead of the path with `LogDir`, because it's already the current directory (see line 293).

eesheesh commented on gapstool.vbs:
```diff
@@ -10,17 +10,20 @@
 Set objShell = WScript.CreateObject("Wscript.Shell")
 Const ForReading = 1, ForWriting = 2, ForAppending = 8
 Const WshRunning = 0, WshFinished = 1, WshFailed = 2
+Const HKEY_LOCAL_MACHINE  = &H80000002 ' From https://msdn.microsoft.com/en-us/library/aa394600(v=vs.85).aspx?cs-lang=vb
```
Two spaces before the comment apostrophe. :)

EmmaStott commented:
Home stretch (I hope!).

2b184792b175da18c90e2cbe1418475bf2f78ea8 fixes the last of the spacing around the constants, contains a ninja patch so that the bypass-list shows as (none) when empty, just like netsh winhttp's output, and no longer tries to write to an absolute path with LogDir.

eesheesh commented:
@EmmaStott Thanks! Can you please squash the commits under one commit (with the correct email address)?

EmmaStott commented:
Squashed into a6e114242427ee344c67b4a70d939b6b6b186285.

eesheesh commented:
Thanks Emma! :)
